File: Meetic.java
package kskills;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class Meetic {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		try {
			BufferedReader r = new BufferedReader(new FileReader("Meetic.txt"));
			
			String line = new String();
			int lineNum = 0;
			
			String num1 = null;
			
			// On lit le fichier
			while ((line = r.readLine()) != null) {
				if (lineNum == 0) {
					num1 = line;
				}
			}
			// Découpe le nombre
			Integer[] numTab = cutNum(num1);
			// Génère la liste des combinaisons possibles
			System.out.println(initListCombi(num1.length()));
			List<String> listCombiIndex = generateAllCombinaisons(initListCombi(num1.length()), num1);
			
			dumpList(listCombiIndex);
			List<String> listCombi = generateListCombi(listCombiIndex, numTab);
			dumpList(listCombi);
			
			System.out.println(findBestNum(listCombi, Integer.parseInt(num1)));
			r.close();
		} catch (FileNotFoundException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}
	
	public static void dumpList(List<String> list){
		System.out.println("");
		System.out.println("Dump:");
		for (String string : list) {
			System.out.println(string);
		}
	}
	
	public static String findBestNum(List<String> listCombiIndex, int num1){
		int bestNum = 0;
		int ecartMin = -num1;
		System.out.println("");
		System.out.println(num1);
		System.out.println("");
		for (String combi : listCombiIndex) {
			int combiNum = Integer.parseInt(combi);
			int ecart = num1-combiNum;
			System.out.println(num1+" "+combiNum+" "+ecart);
			if (ecart < 0 && combiNum != num1 && ecart > ecartMin) {
				ecartMin = ecart;
				bestNum = combiNum;
			}
			
		}
		return ""+bestNum;
	}
	
	/**
	 * Génère la liste des combinaisons de nombres possible à partir de la liste des combinaisons d'index
	 * 
	 * @param listCombiIndex
	 * @param numTab
	 * @return
	 */
	public static List<String> generateListCombi(List<String> listCombiIndex, Integer[] numTab){
		List<String> listeCombi = new ArrayList<String>();
		for (String CombiIndex : listCombiIndex) {
			char[] combiArray = CombiIndex.toCharArray();
			String combi = "";
			for (char c : combiArray) {
				int index = Integer.parseInt(""+c);
				combi += numTab[index];
			}
			listeCombi.add(combi);
		}
		return listeCombi;
	}
	
	/**
	 * initialise la liste des combinaisons
	 * 
	 * @param length
	 * @return
	 */
	public static List<String> initListCombi(int length){
		List<String> listCombi = new ArrayList<String>();
		for (int i=0; i< length; i++) {
			listCombi.add(""+i);
		}
		return listCombi;
	}
	
	/**
	 * Génère la suite de la combinaison
	 * 
	 * @param numPart
	 * @param length
	 * @return
	 */
	public static List<String> generateNext(String numPart, int length){
		List<String> combis = new ArrayList<String>();
		for (int i=0; i<length; i++) {
			if (!numPart.contains(""+i)) {
				String newCombi = numPart;
				newCombi += ""+i;
				combis.add(newCombi);
			}
		}
		return combis;
	}
	
	/**
	 * Génère la liste des combinaisons possibles
	 * 
	 * @param numTab
	 * @return
	 */
	public static List<String> generateAllCombinaisons(List<String> listCombi, String num1){
		List<String> listCombiTemp = new ArrayList<String>();
		if (listCombi.size() > 0) {
			for (int i=0; i<listCombi.size(); i++) {
				List<String> combis = generateNext(listCombi.get(i), num1.length()); 
				listCombiTemp.addAll(combis);
			}
		}
		listCombi.clear();
		listCombi.addAll(listCombiTemp);
		if (listCombi.iterator().next().length() < num1.length()) {
			generateAllCombinaisons(listCombi, num1);
		}
		return listCombi;
	}
	
	/**
	 * Retourne une liste contenant chque chiffre du nombre passé en paramètre
	 * 
	 * @param num
	 * @return
	 */
	public static Integer[] cutNum(String num){
		Integer[] numTab = new Integer[num.length()];
		for (int i=0; i<num.length(); i++) {
			numTab[i] = Integer.parseInt(num.substring(i, i+1));
		}
		return numTab;
	}

}

File: Meetic.php
<?php
/*
 * Vous avez fait une belle rencontre sur Meetic et vous partagez avec votre âme soeur une passion pour le jeu des Chiffres en folie.

	Ce jeu est très simple.
	Le 1er joueur énonce un nombre entier positif N.

	Le 2ème joueur doit arriver en moins de 5 secondes à déterminer quel est le nombre M supérieur à N,  tel que M est le plus proche possible de N,  et que M doit utiliser exactement les mêmes chiffres que N.

	Afin de pouvoir déterminer le 2ème joueur a trouvé la meilleure solution, vous décidez de créer un programme permettant de calculer automatiquement le meilleur score possible.

	 

	 

	Entrée de la fonction
	1 ligne contenant un nombre entier positif N contenant entre 2 et 20 chiffres inclus.

	Sortie de la fonction
	Votre fonction doit retourner un nombre entier M tel que :

	M supérieur à N
	M est composé des mêmes chiffres que N
	M est le plus proche possible de N
 * 
 */

function main()
{
	$stdin = fopen('meetic.txt', 'r');
	$stdout = fopen('php://stdout', 'w');
	
	$num1 = trim(fgets($stdin));
	
	// Découpe le nombre
	$numTab = str_split($num1);
	// Génère la liste des combinaisons d'index possibles
	$listCombiIndex = generateAllCombinaisons(initListCombi(strlen($num1)), $num1);
	// Génère la liste des combinaisons grâce à la liste des combinaisons d'index
	$listCombi = generateListCombi($listCombiIndex, $numTab);
	// On recherche le nombre supérieur le plus proche et on l'affiche
	echo findBestNum($listCombi, intval($num1));
	
	fclose($stdout);
	fclose($stdin);
}

function dump($list){
	echo '<pre>';
	var_dump($list);
	echo '</pre>';
}

	
/**
 * Permet de trouver la meilleur combinaison parmi la liste
 * 
 * @param listCombiIndex
 * @param num1
 * @return
 */
function findBestNum($listCombiIndex, $num1){
	$bestNum = 0;
	// On initialise l'écart minimum avec la valeur négative du nombre d'origine
	$ecartMin = -$num1;
	// On parcours les combinaisons possibles
	foreach ($listCombiIndex as $combi) {
		$combiNum = intval($combi);
		// Si la combinaison est différente du nombre d'origine
		if ($combiNum != $num1) {
			// On calcul l'écart entre la combinaison et le nombre d'origine
			$ecart = $num1-$combiNum;
			// Si l'écart est inférieur à zéro et supérieur à l'écart minimum
			if ($ecart < 0 && $ecart > $ecartMin) {
				$ecartMin = $ecart;
				$bestNum = $combiNum;
			}
		}
		
	}
	return $bestNum;
}

/**
 * Génère la liste des combinaisons de nombres possible à partir de la liste des combinaisons d'index
 * 
 * @param listCombiIndex
 * @param numTab
 * @return
 */
function generateListCombi($listCombiIndex, $numTab){
	$listeCombi = array();
	// On parcourt la liste des combinaisons d'index
	foreach ($listCombiIndex as $CombiIndex) {
		$combiArray = str_split($CombiIndex);
		$combi = "";
		// Pour chaque index, on récupère la valeur correspondante
		foreach ($combiArray as $char) {
			$index = intval($char);
			$combi .= $numTab[$index];
		}
		// On ajoute la combinaison à la liste
		$listeCombi[] = $combi;
	}
	return $listeCombi;
}

/**
 * Initialise la liste des combinaisons d'index
 * 
 * @param length
 * @return
 */
function initListCombi($length){
	$listCombi = array();
	for ($i=0; $i< $length; $i++) {
		$listCombi[] = $i;
	}
	return $listCombi;
}

/**
 * Génère la suite de la combinaison
 * 
 * @param numPart
 * @param length
 * @return
 */
function generateNext($numPart, $length){
	$combis = array();
	// On parcours la liste des indexs
	for ($i=0; $i<$length; $i++) {
		// Si le début de combinaison ne contient pas cet index
		if (strstr(strval($numPart), strval($i)) === false) {
			// On ajoute cette combinaison d'index à la liste
			$combis[] = $numPart . $i;
		}
	}
	return $combis;
}

/**
 * Génère la liste des combinaisons possibles
 * 
 * @param numTab
 * @return
 */
function generateAllCombinaisons($listCombi, $num1){
	$listCombiTemp = array();
	if (sizeof($listCombi) > 0) {
		for ($i=0; $i<sizeof($listCombi); $i++) {
			// On génère la liste des possibilités suivantes pour cette combinaison
			$listCombiNext = generateNext($listCombi[$i], strlen($num1));
			foreach ($listCombiNext as $combi) {
				$listCombiTemp[] = $combi;
			}
		}
	}
	// On vide la liste des combinaisons d'index
	$listCombi = $listCombiTemp;
	// Si la longueur des combinaisons n'est pas égale à la longueur du nombre recherché alors on recommence
	if (strlen($listCombi[0]) < strlen($num1)) {
		$listCombi = generateAllCombinaisons($listCombi, $num1);
	}
	return $listCombi;
}

main();
?>